### PR TITLE
Merge `sway-lib-core` and `sway-lib-std` into this repository

### LINKED ISSUE #830: Consider moving `sway-lib-std` and `sway-lib-core` back into the `sway` repo to make synchronising under CI easier
Currently, we have a cyclic dependency between the `sway` repo and the `sway-lib-core`/`sway-lib-std` repos. `std` and `core` are used throughout the examples and tests under `sway`, and the two libraries implicitly depend on `sway` via the sway language and forc.

We should consider the possibility of moving the `std` and `core` libs back into the `sway` repo to make keeping everything in sync under CI easier and to avoid the need to open two PRs each time we make a change to either the libs or sway.

## forc dependencies in git repo sub-directories

Seeing as Sway doesn't yet have an implicitly included `prelude` through which we can provide `core` and `std`,  we'll need to solve how to refer to the sub-directory of a repo in forc's git depedencies to make this possible. This will be necessary to allow downstream projects to depend on `std` and `core` via git by pointing at the `sway` repo.

Cargo generally seems to "solve" this via their `[workspace]` feature, as crates within subdirectories are almost always a part of some top-level workspace. That said, workspaces are a much broader feature which aims to solve a lot of other problems too, so I'm not sure we should rely on waiting to land a similar feature in forc just to solve this.

An alternative might be to add something like a `root` attribute to forc's git dependency type, where `root` is a relative path from the repo to the subdirectory containing the manifest of the dependency. We could potentially re-use the existing `path` field for this rather than adding a `root` field, as long as we clarify the behaviour that when both `path` and `git` are specified, `path` is relative to the root of the specified git repo rather than the root of the current project.

In the sway repo examples and tests we could solve this by specifying the `core` and `std` dependencies via `path`, however this wouldn't solve anything for downstream projects.

### PR DESCRIPTION
**Pull request by mitchmindtree**
OK! After wading through a long list of blockers I think we might be ready to try and land this :)

---

As discussed in #830, this PR merges the sway `core` and `std` libraries into the `sway` repository.

This doesn't yet address #330, but is one of the final steps towards doing so.

Examples and tests have been updated to depend on `core` and `std` via `path`. This should be useful for ensuring that any changes to either `core` or `std` are accounted for in the E2E tests, while also ensuring that any changes to `sway-core` or `forc` are accounted for in `core` and `std` under CI.

I've added two more dedicated CI jobs for building `sway-lib-core` and `sway-lib-std` so that it should be clear by glancing at a PR if either of those are the cause of CI failure.

At the very least I'll wait to get :no_entry: or :heavy_check_mark:  from @adlerjohn and @nfurfaro as both have pending PRs at `sway-lib-std`, but I've also requested reviews from most folk who touch each of these repos regularly in case anyone else has feedback too!

Closes #830

# History-preserving merge

Rather than copying across the contents of the directories, I've attempted to preserve our history from the `sway-lib-core` and `sway-lib-std` repos by `git merge`ing both into this repository.

To do so, I used the following approach for both:

```
$ git remote add fuellabs/sway-lib-core git@github.com:FuelLabs/sway-lib-core.git
$ git fetch fuellabs/sway-lib-core
$ git merge fuellabs/sway-lib-core/master --allow-unrelated-histories
# Resolve conflicts
```

Conflicts were only related to `README.md`, `.gitignore` and `ci.yml`, but otherwise went smoothly.

# How does `std` depend on `core`?

One issue I noticed when beginning to merge in `std` is that to ensure `std` and `core` are synchronised under each commit, `std` needs to depend on `core` via `path` as there is of course no way to depend on a library via a commit that doesn't exist yet.

I thought `std`'s use of `path` for the `core` dependency might cause issues for projects that would depend on `std` remotely via `git`. However it turns out that `forc` can handle `path` in `dependencies` of `dependencies` without issue **as long as** the dependency's `path` declaration is relative and points to a location within the same git repository.

Intuitively I think this behaviour makes sense and I'm yet to think of a good case against it. That said, I think it should be clearly documented and enforced by `forc` with friendlier errors in a follow-up PR.

# Existing `sway-lib-std` PRs

cc @adlerjohn @nfurfaro I see that there are [a few PRs remaining at sway-lib-std](https://github.com/FuelLabs/sway-lib-std/pulls). Would you like to land those first? Or are you happy to rebase that work onto this repo after this lands? It would be nice if we could avoid too many merge conflicts, but I can manage it fine if you're worried about rebasing the existing work.

# Follow-up

- [ ] Migrate the `./sway-lib-std/tests` into the E2E `stdlib` directory and get them all working.
- [ ] Migrate the `sway-lib-core` and `sway-lib-std` issues and PRs to this repo.

### CODE REVIEW

**Review by sezna — DISMISSED** [review]
This is pretty high-touch, but if i'm not mistaken, the only really _new_ lines of code are the ci job you added? In which case, LGTM, and I love the visibility we get into breaking changes with those separated CI jobs.

**Comment by sezna** [comment]
Do want to wait on John and Nick's thoughts on their PRs though.

**Comment by mitchmindtree** [comment]
> This is pretty high-touch, but if i'm not mistaken, the only really new lines of code are the ci job you added?

Yeah exactly, just the CI jobs and updating the examples and tests to point to `std` and `core` via `path`.

File: .github/workflows/ci.yml
Comment (adlerjohn) [inline .github/workflows/ci.yml:25]: I think we should have a separate `build` job, that this and below depend on. Otherwise it takes 5 minutes to actually build `forc` so these jobs can install.

File: examples/counter/Forc.lock
Comment (adlerjohn) [inline examples/counter/Forc.lock:7]: I'm surprised a path isn't included in the lockfile???

**Comment by adlerjohn** [comment]
I don't care about my open PR(s) for the stdlib, they need `master` anyways.

**Comment by nfurfaro** [comment]
@mitchmindtree I'm fine to rebase my WIP PRs onto the sway repo after this lands.

**Comment by adlerjohn** [comment]
If possible though, could you also include:
- [ ] https://github.com/FuelLabs/sway-lib-std/pull/93

File: examples/counter/Forc.lock
Comment (mitchmindtree) [reply examples/counter/Forc.lock:7]: No `path` is necessary in the lock file, as the local path is declared under `Forc.toml` and there's no ambiguity around which instance of the package to select. This is unlike remote git sources where a single git reference might be associated with a list of possible commits, and we need to be able to reliably fetch the same commit even without control over the remote repository.

File: .github/workflows/ci.yml
Comment (adlerjohn) [reply .github/workflows/ci.yml:25]: >  This would only speed things up if we actually compiled forc as an executable binary and used that in these jobs

These two jobs use `forc` as a binary to execute `forc build` actually!

Comment (Voxelot) [reply .github/workflows/ci.yml:25]: Looks like [cross](https://github.com/cross-rs/cross) could be very useful here. 

ref: https://rust-cli.github.io/book/tutorial/packaging.html#building-binary-releases-on-ci

Comment (mitchmindtree) [reply .github/workflows/ci.yml:25]: Just had a chat with @adlerjohn but adding this for posterity: the `build-sway-lib-core` and `build-sway-lib-std` jobs only took a long time this time as it was the first time they were pushed and built under CI. Consecutive jobs should be closer to the existing `build-sway-examples` time (~2m) for the majority of cases, and seeing as these all run concurrently these should affect the overall time we wait for CI.

**Comment by adlerjohn** [comment]
`assert` was moved, see https://github.com/FuelLabs/sway-lib-std/pull/97. Not sure how this wasn't caught for the stdlib!

**Comment by mitchmindtree** [comment]
> assert was moved, see https://github.com/FuelLabs/sway-lib-std/pull/97. Not sure how this wasn't caught for the stdlib!

No worries, fixing these tests now :) These sorts of issues will be harder let through once it's all merged heh.

**Review by adlerjohn — APPROVED** [review]
Green checkmark means go. Would like @otrho to take a look too.

**Comment by mitchmindtree** [comment]
:clinking_glasses:

**Pull request merged**
